### PR TITLE
Introduce ci_lvms_storage role

### PR DESCRIPTION
**Pull request by fultonj**
The `ci_lvms_storage` role aims to be a drop in replacement for `ci_local_storage`. It uses LVMS (Logical Volume Manager Storage) based on the TopoLVM CSI driver to dynamically provision local storage built from block devices on OCP nodes.

This role requires `cifmw_lvms_disk_list` to contain a list of paths to block devices which already exist on OCP nodes. If the `cifmw_devscripts_config_overrides.vm_extradisks_list` list contains `"vda vdb"`, then it will create these block devices and `cifmw_lvms_disk_list` should be set to `['/dev/vda', '/dev/vdb']`. This role will then pass this list to the `deviceSelector` list used by the `LVMCluster` CRD.

This patch introduces the role but `ci_local_storage` is still used by the `06-deploy-{architecture,edpm}.yml` playbooks unless `cifmw_use_lvms` is true. The `ci_gen_kustomize_values` role has been updated to set the `storageClass` to `lvms-local-storage` only if `cifmw_use_lvms` is true. Otherwise it keeps its default `local-storage`. A similar change has been made to the `test_operator` role so that Tempest uses the chosen storage class.

In a follow up patch we can default `cifmw_use_lvms` to true so that LVMS becomes opt out. For now it is opt in for early testing.

Related: https://issues.redhat.com/browse/OSPRH-2749

As a pull request owner and reviewers, I checked that:
- [x] Appropriate testing is done and actually running
- [x] Appropriate documentation exists and/or is up-to-date:
  - [x] README in the role
  - [x] Content of the docs/source is reflecting the changes

### CODE REVIEW

**Comment by pablintino** [comment]
@fultonj thanks for this contribution, as it's a critical work he are going to need really soon.

**Comment by fultonj** [comment]
1. When `cifmw_lvms=false` (default), automated va1 deploys without an error (using local-storage as usual)
3. When `cifmw_lvms=true`, automated va1 deploys fine but the tempest test is blocked by https://issues.redhat.com/browse/OSPRH-7062 (testing pod requests RWX from LVMS, not an OpenStack on LVMS problem).

Given 1, there is no risk in merging this.

People can then easily test after this merges by setting `cifmw_lvms=true` and we can wait for fix from OSPRH-7062 before we default `cifmw_lvms=true`.

File: roles/ci_gen_kustomize_values/README.md
Comment (cescgina) [inline roles/ci_gen_kustomize_values/README.md:36]: do we have any use-case where we would want to change the prefix? If not we probably do not need to expose this as a parameter and just define it internally when `cifmw_lvms` is true

**Comment by fmount** [comment]
+2 from me

File: roles/ci_lvms_storage/templates/lvms-cluster.yaml.j2
Comment (fultonj) [reply roles/ci_lvms_storage/templates/lvms-cluster.yaml.j2:5]: I've deployed  >1 LVMS cluster on the same OpenShift by calling this role >1 time and overriding the `cifmw_lvms_namespace` each time. However, I didn't test them both at the same time. In that scenario you'd want to override the `cifmw_lvms_disk_list` each time too and it would make sense to override the name `lvmcluster` so I could make that a variable.

File: roles/ci_gen_kustomize_values/README.md
Comment (fultonj) [reply roles/ci_gen_kustomize_values/README.md:36]: We do need `cifmw_ci_gen_kustomize_values_storage_class`. Previously it was not there and `local-storage` was hard-coded in multiple spots.

I'm relying on its prefix variable to always being defined. Do you want to provide another way to solve the problem of getting the right storage class name?

Or do you think I should not document the variable in the readme though I'd keep it in the defaults file? I could prefix it with a `_` to imply that it's private while at the same time keeping it. We could also spend a little time using some other method of getting the name we want. It's mainly there in order to have one flag control different outcomes in multiple roles.

Comment (cescgina) [reply roles/ci_gen_kustomize_values/README.md:36]: I think it's fine just not documenting it in the README if we don't expect users to need to set it. I see why it's needed, and the current approach makes sense to me, I just wanted to clarify if we expected it to have a value different than the default, thanks for the explanation

File: roles/ci_lvms_storage/README.md
Comment (cescgina) [inline roles/ci_lvms_storage/README.md:19]: Should this read `It does not have any tasks` ? I find it confusing otherwise

File: roles/ci_gen_kustomize_values/README.md
Comment (fultonj) [reply roles/ci_gen_kustomize_values/README.md:36]: we do have the prefix pattern already...

- cifmw_ci_gen_kustomize_values_snippets_dir_prefix
- cifmw_ci_gen_kustomize_values_dest_fname_prefix

so I thought it better to be transparent about how it works, especially if someone uses `cifmw_ci_gen_kustomize_values_storage_class` and is surprised to see a prefix there.

File: roles/ci_lvms_storage/README.md
Comment (fultonj) [reply roles/ci_lvms_storage/README.md]: fixed.

File: roles/ci_lvms_storage/templates/lvms-cluster.yaml.j2
Comment (fultonj) [reply roles/ci_lvms_storage/templates/lvms-cluster.yaml.j2]: fixed

https://github.com/openstack-k8s-operators/ci-framework/compare/ce29d5f1c119885a3f605f70d7fc36737417ef97..3054c17e7a4411b885ed79c4948755d56ebf0589

File: roles/ci_gen_kustomize_values/README.md
Comment (cescgina) [reply roles/ci_gen_kustomize_values/README.md:36]: that's a good point about transparency, and the explanation is thorough, I'm not opposed to keep it as it is now, just wanted to clarify

Comment (fultonj) [reply roles/ci_gen_kustomize_values/README.md:36]: https://github.com/openstack-k8s-operators/ci-framework/blob/main/roles/ci_gen_kustomize_values/README.md?plain=1#L23-L28

Comment (fultonj) [reply roles/ci_gen_kustomize_values/README.md:36]: OK, I think if we're going to keep the prefix we should document it for completeness.

**Comment by fultonj** [comment]
> 2. When `cifmw_lvms=true`, automated va1 deploys fine but the tempest test is blocked by https://issues.redhat.com/browse/OSPRH-7062 (testing pod requests RWX from LVMS, not an OpenStack on LVMS problem).
> People can then easily test after this merges by setting `cifmw_lvms=true` and we can wait for fix from [OSPRH-7062](https://issues.redhat.com//browse/OSPRH-7062) before we default `cifmw_lvms=true`.

If you use `quay.io/openstack-k8s-operators/test-operator:latest` then you will not have this issue.

The latest test-operator on quay has fix https://github.com/openstack-k8s-operators/test-operator/pull/96 

I did a fully automated VA1 w/ tempest and LVMS and did not reproduce OSPRH-7062.

```
[zuul@controller-0 ~]$ oc get pvc
NAME                                              STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS         AGE
glance-conversion-glance-default-external-api-0   Bound    pvc-5f1d35f0-ac14-4964-b6f1-148345fe5393   10Gi       RWO            lvms-local-storage   13m
<snip>
persistence-rabbitmq-server-2                     Bound    pvc-26768eed-9f6b-4b6a-bc93-01c003b27720   10Gi       RWO            lvms-local-storage   33m
tempest-tests-38419                               Bound    pvc-52d4ef53-dfe4-47b7-8cde-d5ce929ab271   1Gi        RWO            lvms-local-storage   2m44s
[zuul@controller-0 ~]$ oc get pods | grep tempest
tempest-tests-ddxsr                                               1/1     Running     0             3m36s
[zuul@controller-0 ~]$ oc get sc
NAME                           PROVISIONER   RECLAIMPOLICY   VOLUMEBINDINGMODE      ALLOWVOLUMEEXPANSION   AGE
lvms-local-storage (default)   topolvm.io    Delete          WaitForFirstConsumer   true                   36m
[zuul@controller-0 ~]$ 
```

**Comment by fultonj** [comment]
I recommend we merge this as (per my testing) it doesn't break anything. Then we can ask people to test (including tempest) by setting `cifmw_lvms=true`. Once there is enough comfort we can default `cifmw_lvms` true (for this PR can stay false).

**Review by fmount — COMMENTED** [review]
/lgtm it can be merged and testing can be triggered via `cifmw_lvms=true` until we switch to `LVMS` as the default `storageClass`

File: roles/ci_gen_kustomize_values/README.md
Comment (lewisdenny) [inline roles/ci_gen_kustomize_values/README.md:36]:
```suggestion
* `cifmw_ci_gen_kustomize_values_storage_class_prefix`: (String) Prefix for `storageClass` in generated values.yaml files. Defaults to `"lvms-"` only if `cifmw_lvms` is True, otherwise it defaults to `""`. The prefix is prepended to the `cifmw_ci_gen_kustomize_values_storage_class`. It is not recommended to override this value, instead set `cifmw_lvms` True or False.
```

File: roles/ci_lvms_storage/README.md
Comment (lewisdenny) [inline roles/ci_lvms_storage/README.md:30]: This should be added to `docs/source/usage/01_usage.md`

Comment (lewisdenny) [reply roles/ci_lvms_storage/README.md:30]: [Suggestion] Rather than using a bool for this behavior it would be clearer to use  something like one of these suggestions:
- cifmw_storage_class: "lvms-local-storage" | cifmw_storage_class: "local-storage"
- cifmw_storage_class_provisioner: "ci_lvms_storage" | cifmw_storage_class_provisioner: "ci_local_storage"

File: roles/ci_lvms_storage/defaults/main.yml
Comment (lewisdenny) [inline roles/ci_lvms_storage/defaults/main.yml:25]:
```suggestion
# The "lvms-" prefix is prepended to the cifmw_lvms_storage_class by the lvm-operator
```

File: roles/ci_lvms_storage/files/.gitkeep
Comment (lewisdenny) [inline roles/ci_lvms_storage/files/.gitkeep:1]: [nit] Is this needed?

File: roles/ci_lvms_storage/handlers/main.yml
Comment (lewisdenny) [inline roles/ci_lvms_storage/handlers/main.yml:1]: [nit] Please remove this file

File: roles/ci_lvms_storage/vars/main.yml
Comment (lewisdenny) [inline roles/ci_lvms_storage/vars/main.yml:1]: [nit] Please remove this file

File: roles/test_operator/README.md
Comment (lewisdenny) [inline roles/test_operator/README.md:21]:
```suggestion
* `cifmw_test_operator_storage_class_prefix`: (String) Prefix for `storageClass` in generated Tempest CRD file. Defaults to `"lvms-"` only if `cifmw_lvms` is True, otherwise it defaults to `""`. The prefix is prepended to the `cifmw_test_operator_storage_class`. It is not recommended to override this value, instead set `cifmw_lvms` True or False.
```

**Review by lewisdenny — CHANGES_REQUESTED** [review]
Thank you for the great role and fantastic documentation!

**Comment by lewisdenny** [comment]
/hold
I'm holding this for now while I clarify if we can merge this during beta testing.

@viroel @pablintino @rlandy This doesn't change the default behavior, IMO it's safe but I wanted to confirm.

**Comment by pablintino** [comment]
> /hold I'm holding this for now while I clarify if we can merge this during beta testing.
> 
> @viroel @pablintino @rlandy This doesn't change the default behavior, IMO it's safe but I wanted to confirm.

Hey, this one is in my TODO list, I think it should be safe, but I'd prefer to hold it till @viroel has a full green pipeline for beta.

**Comment by lewisdenny** [comment]
> > /hold I'm holding this for now while I clarify if we can merge this during beta testing.
> > @viroel @pablintino @rlandy This doesn't change the default behavior, IMO it's safe but I wanted to confirm.
> 
> Hey, this one is in my TODO list, I think it should be safe, but I'd prefer to hold it till @viroel has a full green pipeline for beta.

Perfect, glad I held and checked :)

File: roles/test_operator/README.md
Comment (fultonj) [reply roles/test_operator/README.md]: done

File: roles/ci_lvms_storage/handlers/main.yml
Comment (fultonj) [reply roles/ci_lvms_storage/handlers/main.yml]: removed

File: roles/ci_lvms_storage/files/.gitkeep
Comment (fultonj) [reply roles/ci_lvms_storage/files/.gitkeep:1]: removed

File: roles/ci_lvms_storage/defaults/main.yml
Comment (fultonj) [reply roles/ci_lvms_storage/defaults/main.yml]: done.

File: roles/ci_lvms_storage/README.md
Comment (fultonj) [reply roles/ci_lvms_storage/README.md:30]: > This should be added to `docs/source/usage/01_usage.md`

done.

File: roles/ci_gen_kustomize_values/README.md
Comment (fultonj) [reply roles/ci_gen_kustomize_values/README.md]: done

File: roles/ci_lvms_storage/README.md
Comment (fultonj) [reply roles/ci_lvms_storage/README.md:30]: > [Suggestion] Rather than using a bool for this behavior it would be clearer to use something like one of these suggestions:
> 
> * cifmw_storage_class: "lvms-local-storage" | cifmw_storage_class: "local-storage"
> * cifmw_storage_class_provisioner: "ci_lvms_storage" | cifmw_storage_class_provisioner: "ci_local_storage"

It's the LVMS operator which is prepending the prefix. I.e. a prefix of `"lvms-"` is prepended on whatever variable I pass for `cifmw_storage_class` by the LVMs operator. So I default the variable to `local-storage` and then it creates a storage class called `lvms-local-storage`.

  https://github.com/fultonj/ci-framework/blob/ci_lvms_storage/roles/ci_lvms_storage/defaults/main.yml#L26

I do want a single boolean to make it easy for people to use the new feature or not.

Comment (cjeanner) [inline roles/ci_lvms_storage/README.md:30]: hm, late to the party, sorry - but: to match other "boolean" like that, I'd rather go with `cifmw_use_lvms` - that would match `cifmw_use_libvirt`, `cifmw_use_devscripts` and some others we already have through the framework.

Would it make sense? It would require a mighty `sed` in this PR...

**Comment by cjeanner** [comment]
@fultonj care to run `make role_molecule` ? I think some automated steps were missed while creating the new role, leading to a missing "noop" configuration for molecule jobs in CI (needed for job dependency support - and I see I probably was too strict on the files triggering the zuul config check -.-' ).

Thank you!

File: roles/ci_lvms_storage/README.md
Comment (fultonj) [reply roles/ci_lvms_storage/README.md]: I've made the update and we now have boolean variable `cifmw_use_lvms` in place of `cifmw_lvms`

**Comment by cjeanner** [comment]
/approve

This change doesn't change the default behavior - so once we're good with the DS side of the CI (midstream and al) we should be able to merge.

File: roles/ci_lvms_storage/README.md
Comment (fultonj) [reply roles/ci_lvms_storage/README.md]: This actually caused a problem.

I need this var to be in `reproducer-variables.yml` on the controller node it didn't get there because of this.

https://github.com/openstack-k8s-operators/ci-framework/blob/main/roles/reproducer/tasks/configure_controller.yml#L358

Comment (fultonj) [reply roles/ci_lvms_storage/README.md]: I have a plan to fix it tracked in https://issues.redhat.com/browse/OSPRH-7424

**Pull request merged**
